### DISCUSSION TITLE
What's up with the formatting of hs.logger?

### DISCUSSION THREAD

**Discussion by tigersoul925** [Hammerspoon/hammerspoon#3417] [Q&A]
I feel someone probably thought long and hard about the formatting of hs.logger but to me it makes absolutely no sense. I mean, I kind of get it, but also: it's an ugly mess with random passages of spaces. Is there any improved spoon or something I could use to not reinvent the wheel?

I mean look...
![Screenshot 2023-03-08 at 20 54 28](https://user-images.githubusercontent.com/14020174/223831727-c5e3ce54-6dcf-4078-9902-7089920b59a8.png)

**Comment by cmsj** [comment]
You're right, it makes very little sense. I think the aim was to try and prioritise things lining up, but they mostly don't, and the API for the module is kinda over-wraught.

I don't tend to use `hs.logger` very much myself, but if someone would like to work on a refactor to simplify it and improve the output formatting, I would love that.

**Reply by tigersoul925** [reply]
I'd love that too. Unfortunately I don't quite have the time for it right now. Do you use another spoon or module for logging?

**Reply by muescha** [reply]
@cmsj is there a way I can patch this "internal" files on my installation? should I change the file direct on the filesystem or is the best way to copy the file to my `~/.hammerspoon` and then I use it there? But how?

(I look for an easy way - I don't like the long way to compile hammerspoon by myself)

**Reply by latenitefilms** [reply]
You can just copy the hs.logger code from inside the Hammerspoon application bundle into your user folder, and that should override it.

**Reply by muescha** [reply]
I used an extra folder with `logger.lua` and then done a setup with this line:
```lua
hs.logger = require("Helpers.Extensions.logger")
```

which folder I need to get the automatic overwrite?

**Comment by muescha** [comment]
I don't see any problems with the logger. He shows me all the detailed informations. 

Just because of the spaces? I think thats a personal flavour. 

I guess you can tweak it here:
https://github.com/Hammerspoon/hammerspoon/blob/master/extensions/logger/logger.lua

and I see there is an PR to add a logfile handler to the logger:

https://github.com/Hammerspoon/hammerspoon/pull/2627

**Comment by tigersoul925** [comment]
> I don't see any problems with the logger. He shows me all the detailed informations.
> 
> Just because of the spaces? I think thats a personal flavour.
> 
> I guess you can tweak it here: https://github.com/Hammerspoon/hammerspoon/blob/master/extensions/logger/logger.lua
> 
> and I see there is an PR to add a logfile handler to the logger:
> 
> #2627

Yeah, well... if the lines lined up, it would be personal flavor, but they often don't and then there's little point with the layout. But it's not a huge concern. I still use it because it's simple and just works. Would love if someone cleaned up the output though.

**Reply by muescha** [reply]
### setup test

I played with it - I done a copy of logger.lua and required it as new logger:
```lua
hs.logger = require("Helpers.Extensions.logger")
```

### core not using logger

but on many places it is hard to tweak the output, because some libraries don't use the logger internal, they just use `print` statements:
https://github.com/Hammerspoon/hammerspoon/blob/0ccc9d07641a660140d1d2f05b76f682b501a0e8/extensions/window/window.lua#L1026
https://github.com/Hammerspoon/hammerspoon/blob/0ccc9d07641a660140d1d2f05b76f682b501a0e8/extensions/_coresetup/_coresetup.lua#L437
https://github.com/Hammerspoon/hammerspoon/blob/0ccc9d07641a660140d1d2f05b76f682b501a0e8/extensions/_coresetup/_coresetup.lua#L456

### duplicate timestamps

one other thing i see with the logger:
The logger reserve spaces for printing the first timestamp, but there is already a timestamp on the left, so it is duplicate information. (in your screenshot at line: "Init: Identifies ...")

**Reply by tigersoul925** [reply]
I did look at something like this too but it was too much work with too little free time in my case.
